src/loaders/map/map_data.rs

```diff
@@ -0,0 +1,175 @@
+use procedural::*;
+
+pub use super::resource::MapResources;
+use crate::graphics::Color;
+use crate::loaders::map::resource::{LightSettings, WaterSettings};
+use crate::loaders::{ByteConvertable, ByteStream, MajorFirst, Version};
+use crate::world::Tile;
+
+#[derive(ByteConvertable)]
+pub struct MapData {
+    #[version]
+    pub version: Version<MajorFirst>,
+    #[length_hint(40)]
+    pub _ini_file: String,
+    #[length_hint(40)]
+    pub ground_file: String,
+    #[version_equals_or_above(1, 4)]
+    #[length_hint(40)]
+    pub gat_file: Option<String>,
+    #[length_hint(40)]
+    pub _source_file: String,
+    pub water_settings: WaterSettings,
+    pub light_settings: LightSettings,
+    #[version_equals_or_above(1, 6)]
+    pub ground_top: Option<i32>,
+    #[version_equals_or_above(1, 6)]
+    pub ground_bottom: Option<i32>,
+    #[version_equals_or_above(1, 6)]
+    pub ground_left: Option<i32>,
+    #[version_equals_or_above(1, 6)]
+    pub ground_right: Option<i32>,
+    pub resources: MapResources,
+}
+
+#[derive(ByteConvertable)]
+pub struct GatData {
+    #[version]
+    pub version: Version<MajorFirst>,
+    pub map_width: i32,
+    pub map_height: i32,
+    #[repeating(self.map_width * self.map_height)]
+    pub tiles: Vec<Tile>,
+}
+
+#[derive(ByteConvertable)]
+pub struct GroundData {
+    #[version]
+    pub version: Version<MajorFirst>,
+    pub width: i32,
+    pub height: i32,
+    pub zoom: f32,
+    pub texture_count: i32,
+    pub texture_name_length: i32,
+    #[repeating(self.texture_count)]
+    #[length_hint(self.texture_name_length)]
+    pub textures: Vec<String>,
+    pub light_map_count: i32,
+    pub light_map_width: i32,
+    pub light_map_height: i32,
+    pub light_map_cells_per_grid: i32,
+
```

Comment:
You should run `cargo fmt --all` to get rid of these empty lines

Reply:
not really, I ran fmt --all

Reply:
Yes that is my bad, I thought that `cargo fmt` removes these empty lines but apparently it does not. It would be great if you could remove them manually, alternatively I would also be fine with a comment describing the fields below instead of an empty line.